crates/archive/src/writer.rs

```diff
@@ -1,111 +1,82 @@
-use crate::chain_builder::{AnyChainBuilder, ChainBuilderBox};
-use parking_lot::Mutex;
+use crate::fs::FSRef;
+use crate::layout::DataChunk;
+use crate::metrics;
 use parquet::arrow::ArrowWriter;
 use parquet::basic::{Compression, ZstdLevel};
 use parquet::file::properties::WriterProperties;
-use sqd_data_core::ChunkProcessor;
-use sqd_dataset::TableDescription;
+use rayon::prelude::*;
+use sqd_data_core::PreparedChunk;
+use sqd_dataset::{DatasetDescriptionRef, TableDescription};
 use std::fs::File;
-use std::ops::DerefMut;
 use std::path::Path;
+use tokio::sync::mpsc::UnboundedReceiver;
 
 
-struct State {
-    builder: ChainBuilderBox,
-    processor: ChunkProcessor
+pub struct WriterItem {
+    pub chunk: DataChunk,
+    pub data: PreparedChunk,
+    pub description: DatasetDescriptionRef,
 }
 
 
-impl State {
-    fn spill_on_disk(&mut self) -> anyhow::Result<()> {
-        self.builder.chunk_builder().submit_to_processor(&mut self.processor)?;
-        self.builder.chunk_builder_mut().clear();
-        Ok(())
-    }
+pub struct Writer {
+    fs: FSRef,
+    chunk_receiver: UnboundedReceiver<WriterItem>,
 }
 
 
-pub struct ParquetWriter {
-    state: Mutex<State>,
-    spare_processor: Mutex<Option<ChunkProcessor>>,
-    memory_threshold: usize,
-    max_num_rows: usize
-}
-
-
-impl ParquetWriter {
-    pub fn new(chain_builder: ChainBuilderBox) -> ParquetWriter {
-        ParquetWriter {
-            state: Mutex::new(State {
-                processor: chain_builder.chunk_builder().new_chunk_processor(),
-                builder: chain_builder
-            }),
-            spare_processor: Mutex::new(None),
-            memory_threshold: 40 * 1024 * 1024,
-            max_num_rows: 200_000
-        }
-    }
-
-    pub fn push(&self, block_json: &[u8]) -> anyhow::Result<()> {
-        let mut state = self.state.lock();
-        
-        state.builder.push(block_json)?;
-        
-        if state.builder.chunk_builder().byte_size() > self.memory_threshold {
-            state.spill_on_disk()?;
-        }
-        Ok(())
+impl Writer {
+    pub fn new(fs: FSRef, chunk_receiver: UnboundedReceiver<WriterItem>) -> Writer {
+        Writer { fs, chunk_receiver }
     }
 
-    pub fn last_parent_block_hash(&self) -> &str {
-        todo!()
-    }
+    pub async fn start(&mut self) -> anyhow::Result<()> {
+        while let Some(mut item) = self.chunk_receiver.recv().await {
+            tracing::info!("writing {}", item.chunk.path());
 
-    pub fn buffered_bytes(&self) -> usize {
-        let state = self.state.lock();
-        state.builder.chunk_builder().byte_size() + state.processor.byte_size()
-    }
+            let last_block = item.chunk.last_block;
+            let target_dir = tempfile::tempdir()?.into_path();
```

Comment:
It's a bit strange that a generic `FSRef` is passed to the `Writer`, but it still uses local fs in this method

Reply:
i believe it's parquet format limitation. we should write data to a local file first and then we can upload it to s3